File: 2016/Day12/Day12.pl
use strict;
use warnings;

use Data::Dumper;

my $inputfile = shift;
if (!defined($inputfile)) {
  $inputfile = 'Sample.txt';
}

open(my $fh, '<', $inputfile) or die "Couldn't open $inputfile: $!";
my @commands = ();
foreach my $line (<$fh>) {
  $line =~ s/\R//g; # Trim newlines
  if ($line =~ m/^cpy (\d+|a|b|c|d) (a|b|c|d)$/) {
    push(@commands, ['cpy', $1, $2]);
  }
  elsif ($line =~ m/^jnz (\d+|a|b|c|d) (-?)(\d+)$/) {
    my ($reg, $neg, $offset) = ($1, $2, int($3));
    if (defined($neg) && $neg eq '-') {
      $offset = -1 * $offset;
    }
    push(@commands, ['jnz', $reg, $offset]);
  }
  elsif ($line =~ m/^inc (a|b|c|d)$/) {
    push(@commands, ['inc', $1]);
  }
  elsif ($line =~ m/^dec (a|b|c|d)$/) {
    push(@commands, ['dec', $1]);
  }
  else {
    die "Unrecognized input sequence: '$line'";
  }
}
close($fh);

#print "Commands:\n";
#print Dumper \@commands;

my $numcommands = scalar(@commands);
my ($a, $b, $c, $d) = (0, 0, 1, 0); # Registers
my $i = 0; # Instruction pointer

while (0 <= $i && $i < $numcommands) {
  my $cmd = $commands[$i][0];

  if ($cmd eq 'cpy') {
    my $val = $commands[$i][1];
    if ($val eq 'a') {
      $val = $a;
    } elsif ($val eq 'b') {
      $val = $b;
    } elsif ($val eq 'c') {
      $val = $c;
    } elsif ($val eq 'd') {
      $val = $d;
    } else {
      $val = int($val);
    }

    my $dest = $commands[$i][2];
    if ($dest eq 'a') {
      $a = $val;
    } elsif ($dest eq 'b') {
      $b = $val;
    } elsif ($dest eq 'c') {
      $c = $val;
    } else {
      $d = $val;
    }

    $i = $i + 1;
  }
  elsif ($cmd eq 'jnz') {
    my ($reg, $offset) = ($commands[$i][1], $commands[$i][2]);

    if (($reg eq 'a' && $a != 0)
      || ($reg eq 'b' && $b != 0)
      || ($reg eq 'c' && $c != 0)
      || ($reg eq 'd' && $d != 0)) {

      $i = $i + $offset;
    }
    elsif (int($reg) != 0) {
      $i = $i + $offset;
    }
    else {
      $i = $i + 1;
    }
  }
  elsif ($cmd eq 'inc') {
    my $reg = $commands[$i][1];
    if ($reg eq 'a') {
      $a = $a + 1;
    }
    elsif ($reg eq 'b') {
      $b = $b + 1;
    }
    elsif ($reg eq 'c') {
      $c = $c + 1;
    }
    else {
      $d = $d + 1;
    }

    $i = $i + 1;
  }
  else {
    # cmd is dec
    my $reg = $commands[$i][1];
    if ($reg eq 'a') {
      $a = $a - 1;
    }
    elsif ($reg eq 'b') {
      $b = $b - 1;
    }
    elsif ($reg eq 'c') {
      $c = $c - 1;
    }
    else {
      $d = $d - 1;
    }

    $i = $i + 1;
  }
}

print "Finished execution: state is:\n";
print "\ta: $a\n";
print "\tb: $b\n";
print "\tc: $c\n";
print "\td: $d\n";

File: 2016/Day12/Day12.cpp
// Day12.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

enum CommandType {
   COMMAND_COPY,
   COMMAND_INCREMENT,
   COMMAND_DECREMENT,
   COMMAND_JUMP_NOT_ZERO
};

struct Command {
   CommandType type;
   std::wstring value1;
   std::wstring value2;
};

std::vector<Command> GetCommands(const std::wstring& inputFile) {
   std::vector<Command> commands;
   std::wifstream input{ inputFile };
   if (!input.is_open()) {
      return commands;
   }

   std::wstring token;
   input >> token;
   while (input.good()) {
      Command cmd;
      if (token == L"cpy") {
         cmd.type = COMMAND_COPY;
         input >> cmd.value1;
         input >> cmd.value2;
      }
      else if (token == L"inc") {
         cmd.type = COMMAND_INCREMENT;
         input >> cmd.value1;
      }
      else if (token == L"dec") {
         cmd.type = COMMAND_DECREMENT;
         input >> cmd.value1;
      }
      else if (token == L"jnz") {
         cmd.type = COMMAND_JUMP_NOT_ZERO;
         input >> cmd.value1;
         input >> cmd.value2;
      }
      else {
         std::wcerr << L"Unrecognized command: " << token << std::endl;
         throw L"Unrecognized command";
      }

      commands.push_back(cmd);
      input >> token;
   }

   return commands;
}

int RunCommands(const std::vector<Command>& commands, int a, int b, int c, int d) {
   std::wcout << L"Initial state: a = " << a << L", b = " << b << L", c = " << c << L", d = " << d << std::endl;

   size_t instruction = 0;
   while (0 <= instruction && instruction < commands.size()) {
      Command currentCommand = commands[instruction];
      int val;
      switch (currentCommand.type) {
      case COMMAND_COPY:
         // value1 is source register or value, value2 is destination register
         if (currentCommand.value1 == L"a") {
            val = a;
         }
         else if (currentCommand.value1 == L"b") {
            val = b;
         }
         else if (currentCommand.value1 == L"c") {
            val = c;
         }
         else if (currentCommand.value1 == L"d") {
            val = d;
         }
         else {
            val = _wtoi(currentCommand.value1.c_str());
         }

         if (currentCommand.value2 == L"a") {
            a = val;
         }
         else if (currentCommand.value2 == L"b") {
            b = val;
         }
         else if (currentCommand.value2 == L"c") {
            c = val;
         }
         else {
            d = val;
         }

         ++instruction;
         break;
      case COMMAND_DECREMENT:
         // value1 is register, value2 is unused
         if (currentCommand.value1 == L"a") {
            --a;
         }
         else if (currentCommand.value1 == L"b") {
            --b;
         }
         else if (currentCommand.value1 == L"c") {
            --c;
         }
         else {
            --d;
         }

         ++instruction;
         break;
      case COMMAND_INCREMENT:
         // value1 is register, value2 is unused
         if (currentCommand.value1 == L"a") {
            ++a;
         }
         else if (currentCommand.value1 == L"b") {
            ++b;
         }
         else if (currentCommand.value1 == L"c") {
            ++c;
         }
         else {
            ++d;
         }

         ++instruction;
         break;
      case COMMAND_JUMP_NOT_ZERO:
         // value1 is register or value, value 2 is offset
         if (currentCommand.value1 == L"a") {
            val = a;
         }
         else if (currentCommand.value1 == L"b") {
            val = b;
         }
         else if (currentCommand.value1 == L"c") {
            val = c;
         }
         else if (currentCommand.value1 == L"d") {
            val = d;
         }
         else {
            val = _wtoi(currentCommand.value1.c_str());
         }

         if (val == 0) {
            // No offset
            ++instruction;
         }
         else {
            // Offset
            int offset = _wtoi(currentCommand.value2.c_str());
            instruction += offset;
         }
         break;
      }
   }

   std::wcout << L"Final state: a = " << a << L", b = " << b << L", c = " << c << L", d = " << d << std::endl;

   return a;
}

int wmain(int argc, wchar_t *argv[]) {
   std::wstring inputFile;
   if (argc == 2) {
      inputFile = argv[1];
   }
   else {
      inputFile = L"Sample.txt";
   }

   std::vector<Command> commands = GetCommands(inputFile);

   
   
   // Part 1
   int answer1{ RunCommands(commands, 0, 0, 0, 0) };

   // Part 2
   int answer2{ RunCommands(commands, 0, 0, 1, 0) };

   std::wcout << L"Answers: part 1 " << answer1 << L", part 2 " << answer2 << std::endl;

   return 0;
}
